### PR TITLE
Unnumbered BGP Peering design document

### PR DESCRIPTION
**Pull request by karampok**
/kind design

**What this PR does / why we need it**:

**Special notes for your reviewer**:

In part of this PR there is
1. A design document about unnumbered BGP peering, discussing implementation and API changes
2. A draft version of how an e2e test can run, more for discussion rather than actually merging that file in that PR.

My plan would be that first we discuss how the raw frr config would look like, then the API in frr-k8s, then in metallb.

Thanks

**Release note**:

```release-note
None
```

### CODE REVIEW

File: design/unnumbered-bgp.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+    %% Define styles for subgraphs and nodes
+    classDef node fill:#f0f0f0,stroke:#000000,stroke-width:1px;
+    class Node0 node;
+    class Node1 node;
+```
+
+## Design details
+
+The target `show running-config` in the FRR process running in the k8s node
```
Comment (oribon) [inline design/unnumbered-bgp.md:161]: can you make this more alike the config we have today for the speakers? also, can you add a few words about the need for peergroups (iirc it's because the usual directives do not work with the interface's name)

File: design/unnumbered-bgp.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+Peer groups 1, using 64 bytes of memory
+
+Neighbor                  V         AS   MsgRcvd   MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd   PfxSnt Desc
+unnumbered-p2p-peer(net0) 4      65004       512       512        1    0    0 00:25:15            0        1 TOR
+
+Total number of neighbors 1
+```
+### Topics
+
+#### Remote-AS
```
Comment (oribon) [inline design/unnumbered-bgp.md:310]: not really sure how this relates here (it says here we can specify internal/external/asn which fits with any model), is there something special about this in regards to unnumbered?

File: design/unnumbered-bgp.md [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+#### frr-k8s side
+
+```go
+...
+
+type Neighbor struct {
+  Description string `json:"description"` // optional
+  Address string `json:"address"` // must be empty if interface != nil
+  Interfaces []string `json:"interface"` //
```
Comment (oribon) [inline design/unnumbered-bgp.md:352]: why is this an array?

File: design/unnumbered-bgp.md [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+### The API Extension
+
+#### frr-k8s side
+
+```go
+...
+
+type Neighbor struct {
+  Description string `json:"description"` // optional
+  Address string `json:"address"` // must be empty if interface != nil
```
Comment (oribon) [inline design/unnumbered-bgp.md:351]: does this really need to be empty? could both being specified solve anything for https://github.com/metallb/metallb/issues/2423?

File: design/unnumbered-bgp.md [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+  Interfaces []string `json:"interface"` //  
+  PeerGroup string `json:"peerGroup"` // optional, otherwise value <interface>-peergroup
```
Comment (oribon) [inline design/unnumbered-bgp.md:353]: is there a reason for exposing this to the user (in particular as part of "unnumbered")?

File: design/unnumbered-bgp.md [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+
+#### BGP RouterID
+
+//TODO is there a requirement for that?
+
+### Traffic forwarding
+
+FRR advertises IPv4 prefixes with IPv6 next hops using encoding specified in RFC 5549 (now RFC 8950).
+
+### The API Extension
```
Comment (oribon) [inline design/unnumbered-bgp.md:342]: nit: I think it would be clearer to say which fields you think should be added/modified, instead of the complete struct as it is today

File: design/unnumbered-bgp.md [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+
+```go
+...
+
+type Neighbor struct {
+  Description string `json:"description"` // optional
+  Address string `json:"address"` // must be empty if interface != nil
+  Interfaces []string `json:"interface"` //  
+  PeerGroup string `json:"peerGroup"` // optional, otherwise value <interface>-peergroup
+  RemoteAS string `json:"remoteAS"` // it can be "ASN number or internal or external [or auto]"
```
Comment (oribon) [inline design/unnumbered-bgp.md:354]: like the other comment, not sure this is an "unnumbered" thing

File: design/unnumbered-bgp.md [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+```
+
+#### MetalLB
+
+```go
+...
+type BGPPeerSpec struct {
+	MyASN uint32 `json:"myASN"`
+	ASN uint32 `json:"peerASN"` // or internal or external
+	Address string `json:"peerAddress"` // or interface name e.g. net0
```
Comment (oribon) [inline design/unnumbered-bgp.md:375]: what's the reason for address/interface being separate in frr-k8s but not here?

File: design/unnumbered-bgp.md [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+### User Stories
+
+There is direct link interface between the node (FRR process acting as BGP peer) and the
+ToR (acting as an BGP router).
+
+
+#### Story 1 - Unnumbered {i,E}BGP on primary node interface
+
+```mermaid
```
Comment (oribon) [inline design/unnumbered-bgp.md:52]: neat!

File: e2etest/bgptests/unnumber.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+var _ = ginkgo.FDescribe("FRRK8S-MODE DUALSTACK - Unnumbered BGP", func() {
+	defer ginkgo.GinkgoRecover()
+	var (
+		testNamespace string
+		peers         []*frrcontainer.FRR
+	)
+
+	ginkgo.BeforeEach(func() {
+		c := frrcontainer.Config{
```
Comment (oribon) [inline e2etest/bgptests/unnumber.go:40]: the goal is to have this as part of the infra setup to make sure everything works with unnumbered?

File: design/unnumbered-bgp.md [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+
+
+### Testing
+
+#### E2E tests
+
+- Modify the testing infra to include point-to-point links which means
+connect containers using VETH, and not through a bridge. We need NET-ADMIN cap (or sudo) when creating the VETH
+- Create/Delete peer per specific test not to take resources from runners outside the runtime of the test
+- This peer needs different configuration (not IPs, interface names), keep a raw config
```
Comment (oribon) [inline design/unnumbered-bgp.md:426]: iiuc this will all be covered in the initial implementation?

File: design/unnumbered-bgp.md [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+- There is no straightforward way to test unnumbered in the primary node interface. The testing setup (kind creation)
```
Comment (oribon) [inline design/unnumbered-bgp.md:427]: I wonder if there's some value in adding another worker+container that are peered using unnumbered (no p2p, but there will be no conflicts as they're the only routers in the network using it) to get some feedback on peering with the primary node interface? wdyt?

**Review by oribon — COMMENTED** [review]
thanks! great work, did a first pass

File: design/unnumbered-bgp.md (hunk 1 above)
Comment (karampok) [reply design/unnumbered-bgp.md:161]: I will go deeper on what we have today in the speaker, that part if not very clear.

About PeerGroups:
1. Peergroups is the only way I managed to make it working, and all reference/examples did make use of them. So I observe we need them but I did not find doc.
2. I think the case of having two ToRs switch will be a needed, so I am thinking that we can have PG with two interfaces assigned.

File: design/unnumbered-bgp.md (hunk 2 above)
Comment (karampok) [reply design/unnumbered-bgp.md:310]: I thought that unnumber BGP referred to both interface without values, and ASN without values, but not be the case.

What only confuses me is that 
[neighbor PEER interface remote-as <internal|external|auto|ASN>](https://docs.frrouting.org/en/latest/bgp.html#clicmd-neighbor-PEER-interface-remote-as-internal-external-auto-ASN)

which  talks about unnumber, but when I added this directive  did not work  as expected. in my lab.

I would suggest that this doc documents the API changes for that also, but I will not work on the implementation.

File: design/unnumbered-bgp.md (hunk 3 above)
Comment (karampok) [reply design/unnumbered-bgp.md:352]: To support the dual ToRs case.

File: design/unnumbered-bgp.md (hunk 8 above)
Comment (karampok) [reply design/unnumbered-bgp.md:375]: Not really, it could be hidden and taken an internal value (if we decided using peergroups)

File: e2etest/bgptests/unnumber.go (hunk 10 above)
Comment (karampok) [reply e2etest/bgptests/unnumber.go:40]: My preference would be to create the new container the unnumber peering as part of the test, to avoid having a container running for the full rutime of the tests. Do you see a reason to move it to the infra setup?

File: design/unnumbered-bgp.md (hunk 11 above)
Comment (karampok) [reply design/unnumbered-bgp.md:426]: Can you elaborate what you mean?

File: design/unnumbered-bgp.md (hunk 12 above)
Comment (karampok) [reply design/unnumbered-bgp.md:427]: Do you mean to have a single worker as speaker, and have single external container frr connected on the same docker network bridge? 

My concern is that locally the linux bridge does not forwarding RA when LLA address is the source (at least in my fedora laptop).

For sure we need to verifiy how that works on primary interface, and especially when the interface has also a non-LLA address configured.

File: design/unnumbered-bgp.md (hunk 5 above)
Comment (karampok) [reply design/unnumbered-bgp.md:353]: Not really, it could be hidden and taken an internal value (if we decided using peergroups)

File: design/unnumbered-bgp.md (hunk 1 above)
Comment (oribon) [reply design/unnumbered-bgp.md:161]: > Peergroups is the only way I managed to make it working, and all reference/examples did make use of them. So I observe we need them but I did not find doc.

right, this is what I meant by "the usual directives do not work with the interface's name" :) let's add this to this section so it's clear why we suddenly need peer groups.

> I think the case of having two ToRs switch will be a needed, so I am thinking that we can have PG with two interfaces assigned.

I see, but I guess this could be achieved by just having 2 different peers?

File: design/unnumbered-bgp.md (hunk 2 above)
Comment (oribon) [reply design/unnumbered-bgp.md:310]: >  would suggest that this doc documents the API changes for that also

I think it would be better to keep this focused on unnumbered, when we get to implementing that feature (either if it's before or after this is implemented) we can see how they work together.

File: design/unnumbered-bgp.md (hunk 5 above)
Comment (oribon) [reply design/unnumbered-bgp.md:353]: so I think we should drop this for now so the doc stays focused on unnumbered, when/if we decide peer groups are valuable as part of the api we'll tackle them separately :)

File: e2etest/bgptests/unnumber.go (hunk 10 above)
Comment (oribon) [reply e2etest/bgptests/unnumber.go:40]: > Do you see a reason to move it to the infra setup?

to run all the existing tests against the "unnumbered" scenario (unless we decide we're good with having only a single test for unnumbered)

File: design/unnumbered-bgp.md (hunk 11 above)
Comment (oribon) [reply design/unnumbered-bgp.md:426]: asking if these points will be covered or there's still a question about them, while the last point (about primary interfaces) still needs to be thought of

File: design/unnumbered-bgp.md (hunk 12 above)
Comment (oribon) [reply design/unnumbered-bgp.md:427]: > My concern is that locally the linux bridge does not forwarding RA when LLA address is the source (at least in my fedora laptop).

oh, didn't you run into the "random" peering we talked about?

> Do you mean to have a single worker as speaker, and have single external container frr connected on the same docker network bridge?

right (adding another worker / designating one), then the "randomness" will be okay - since there are only 2 peers on the same network using unnumbered they'll always peer with each other

File: design/unnumbered-bgp.md (hunk 1 above)
Comment (karampok) [reply design/unnumbered-bgp.md:161]: ok, I will go through to see how two neighbors would look like, and check again if Peer Group can be avoided

File: design/unnumbered-bgp.md (hunk 2 above)
Comment (karampok) [reply design/unnumbered-bgp.md:310]: sounds good, I will remove that section

File: design/unnumbered-bgp.md (hunk 5 above)
Comment (karampok) [reply design/unnumbered-bgp.md:353]: sounds good, I will not add any public API about PG

File: e2etest/bgptests/unnumber.go (hunk 10 above)
Comment (karampok) [reply e2etest/bgptests/unnumber.go:40]: I see your point, but "all existing" means double the time and mabye adding complexity in the setupinfra. But no strong opinion, we can come back to that  on a later step.

Comment (oribon) [reply e2etest/bgptests/unnumber.go:40]: > means double the time

not sure about that, it would just be another container as part of the tests

> mabye adding complexity

probably, this pr is a good place to understand if it's reasonable or not :)

File: design/unnumbered-bgp.md [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+Linux sends and receives IPv4 packets over an interface that does not have an IPv4 address.
+It does not look we need to add address in the loopback.
+
+#### RA enabled on both ends of the link
+
+According to the design of unnumber BGP peering, we need to enable RAs.
+
+#### BGP RouterID
+
+//TODO is there a requirement for that?
```
Comment (oribon) [inline design/unnumbered-bgp.md:317]: wdym here?

File: design/unnumbered-bgp.md [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+Unnumbered BGP utilizes IPv6 link local address to automatically decide who to
+connect.
+
+## Limitation/Requirements
+
+- To use unnumbered BGP feature, there must be a link (point-to-point, directly
+connected) between the interfaces two BGP daemons are using
+- IPv6 must be enabled on that interfaces, and in specific LLA (link Local IPv6 address) enabled
+(but the use of IPv6 LLA does not require a full IPv6 deployment)
+- RAs must be enabled (it should be safe when p2p and no IP address on the interface TODO)
```
Comment (oribon) [inline design/unnumbered-bgp.md:24]: is there still a TODO here?

File: design/unnumbered-bgp.md [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+
+#### Option 1. Re-Use Address field
+
+We can re-use `Address string` to either be IP or interface, if not valid IP we
+indicate Unnumbered BGP peering. This looks the minimal API, but it will not
+have clean error path. E.g. if the user input is not valid IP, the
+implementation will shadow all the existing errors and will fail in the
+runtime.
+
+#### Option 2. Add a new Interface field
```
Comment (oribon) [inline design/unnumbered-bgp.md:362]: I think this is the most valid one, and one that matches FRR the most in a sense (since it seems there's a distinction between interfaces and addresses) without locking us. Also, the docs don't need to be as complicated (stating all of the cases) as it's simply following the what frr is doing, that is we can just shortly say, if address is not specified the peering happens over the interface (unnumbered), if address is specified it is assumed the it is lla.

File: design/unnumbered-bgp.md [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+// == extra FRR directive `neighbor fe80::a876:4dff:fe77:408 interface eth0`
+// 4. If address is LLA and empty interface, then fail
+// 5. If address is empty and interface has value then the BGP unnumber peering config will take place
+// ==  modify FRR directive `neighbor PEER interface remote-as <ASN>`
+// && replace any address directive to use interface e.g.  `neighbor net0 timers 180 540`
+// && enable RAs on this interface
+// 6. If address has value (non-LLA) and interface non empty, then fail
+```
+That option might be not ideal (subjective opinion) because documentation looks complicated,
+and user might have to do trial-and-error approach. Plus user needs to understand corner cases.
```
Comment (oribon) [inline design/unnumbered-bgp.md:382]: the user will need to understand the usecases as it does for frr, is there extra complexity beyond that?

File: design/unnumbered-bgp.md [hunk 17]
```diff
@@ -0,0 +1,498 @@
...
+}
+
+```
+
+##### MetalLB
+
+```go
+type BGPPeerSpec struct {
+  Address string `json:"peerAddress"`
+  Interface bool `json:"interface"`
```
Comment (oribon) [inline design/unnumbered-bgp.md:401]: I guess this should be string?

File: design/unnumbered-bgp.md [hunk 18]
```diff
@@ -0,0 +1,498 @@
...
+```yaml
+apiVersion: metallb.io/v1beta2
+kind: BGPPeer
+metadata:
+  name: peer-unnumber
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  ASN: 645000
+  peerAddress: "" // empty, can be omitted
```
Comment (oribon) [inline design/unnumbered-bgp.md:417]: just omit here for the example

File: design/unnumbered-bgp.md [hunk 19]
```diff
@@ -0,0 +1,498 @@
...
+
+#### E2E tests
+
+- Modify the testing infra to include point-to-point links which means connect
+containers using VETH, and not through a bridge. We need NET-ADMIN cap (or sudo)
+when creating the VETH.
+
+- Create/Delete peer per that specific test not to take resources from runners
+outside the runtime of the test. This peer needs different configuration (not
+IPs and interface names), keep separate raw configuration.
```
Comment (oribon) [inline design/unnumbered-bgp.md:483]: not sure I understand, you're proposing a separate test for the unnumbered case or expanding the existing containers setup?

Comment (karampok) [reply design/unnumbered-bgp.md:476]: I propose a separate isolated test for that use case.
(but no strong argument against expanding container setup).
Do you have a preference here?

File: design/unnumbered-bgp.md (hunk 18 above)
Comment (karampok) [reply design/unnumbered-bgp.md:416]: ok, will do

File: design/unnumbered-bgp.md (hunk 17 above)
Comment (karampok) [reply design/unnumbered-bgp.md:400]: yes, will fix

File: design/unnumbered-bgp.md [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+  no bgp network import-check
+
+  neighbor net0 interface remote-as 65004
+  neighbor net0 description TOR
+  neighbor net0 port 179
+  neighbor net0 bfd
+  neighbor net0 bfd profile fornet0
+  neighbor net0 timers 180 540
+  neighbor net0 timers connect 30
+  neighbor net0 graceful-restart
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:175]: I think I did, but I will double check, is important

File: design/unnumbered-bgp.md [hunk 21]
```diff
@@ -0,0 +1,498 @@
...
+
+## Design details
+
+The `show running-config` in the FRR process running in the k8s node
+looks like
+
+
+```
+interface net0
+  ipv6 nd ra-interval 10
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md]: As it seems https://github.com/FRRouting/frr/commit/4a04e5f7968d46e446bdb530724da18b90b7f5f8 we do not need at all the interface block, it works without as RA auto-enabled in the unnumber case. (and I do not see any reason to change RA interval in a p2p link)

File: design/unnumbered-bgp.md [hunk 22]
```diff
@@ -0,0 +1,498 @@
...
+
+        end
+
+        %% Top Right Subgraph
+        subgraph Node1
+            direction TB
+            B1 --- B2(cni)
+            B2 --- B3
+            subgraph Pod1[pod]
+              B3[eth0]
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:71]: I wanted to point out that there a pod/cni part, and a BGP process that is in th host. I should add the BGP process in the first diagram. I think there is no harm to have it

File: design/unnumbered-bgp.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (oribon) [reply design/unnumbered-bgp.md]: 👍 (saw that you removed it completely)

File: design/unnumbered-bgp.md (hunk 19 above)
Comment (oribon) [reply design/unnumbered-bgp.md:476]: I'd say if it's possible to have all the tests ran on unnumbered in a clean way it's preferable

File: design/unnumbered-bgp.md (hunk 14 above)
Comment (karampok) [reply design/unnumbered-bgp.md]: no todo, removed

File: design/unnumbered-bgp.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:174]: looks good, updated the doc and did my smoke test using route-maps and interface

File: design/unnumbered-bgp.md [hunk 23]
```diff
@@ -0,0 +1,498 @@
...
+  address-family ipv4 unicast
+    network 5.5.5.5/32
+    neighbor net0 activate
+ exit-address-family
+```
+
+* keep the format as in [neighborsession.tmpl](internal/bgp/frr/templates/neighborsession.tmpl)
+* instead of address (ip) to be interface name
+* enable RAs in that interface name
+* must exclude FRR directives that do not accept interface e.g.
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:205]: I found that this directive `disable-connected-check` can be combined with inteface. So it must be excluded otherwise throughs errors. I have not find other directive

File: design/unnumbered-bgp.md [hunk 24]
```diff
@@ -0,0 +1,498 @@
...
+    exit-address-family
+    !
+    address-family ipv6 unicast
+        redistribute connected
+        neighbor MTLB activate
+    exit-address-family
+    exit
+```
+
+An example ToR configuration when based on Junos
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:250]: imo yes, maybe not our task, but unless I see that real config in real setup, I am not sure our use cases are valid.

File: design/unnumbered-bgp.md [hunk 25]
```diff
@@ -0,0 +1,498 @@
...
+
+```
+//TODO
+```
+
+###  How a peering will look like
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:256]: I find it nice when looking a such design doc, to get a feeling how a working setup looks like. Do you want me to remove it?

File: design/unnumbered-bgp.md (hunk 13 above)
Comment (karampok) [reply design/unnumbered-bgp.md:331]: I have a lack of  BGP knowledge what is the general impact of this RouterID in our use cases?  Do you know?

File: design/unnumbered-bgp.md (hunk 16 above)
Comment (karampok) [reply design/unnumbered-bgp.md:396]: imo yes :)

Comment (oribon) [reply design/unnumbered-bgp.md:396]: so I think it should be more detailed, because it is not really clear atm

File: design/unnumbered-bgp.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (oribon) [reply design/unnumbered-bgp.md:205]: not sure I follow, are there options we can't use for unnumbered or it just needs a different configuration option?

File: design/unnumbered-bgp.md [hunk 26]
```diff
@@ -0,0 +1,508 @@
...
+### Topics
+
+#### Loopback IPv4 address
+
+Linux sends and receives IPv4 packets over an interface that does not have an IPv4 address.
+It does not look we need to add address in the loopback.
+
+#### RA enabled on both ends of the link
+
+According to the design of unnumber BGP peering, we need to enable RAs. By
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:323]: can you expand on what RA is?

File: design/unnumbered-bgp.md (hunk 13 above)
Comment (fedepaol) [reply design/unnumbered-bgp.md:331]: frr sets the router id with the lowest ip of the interfaces. So far, we didn't have a need to do otherwise. If this is not relevant please remove the section as it's distracting.

File: design/unnumbered-bgp.md [hunk 27]
```diff
@@ -0,0 +1,508 @@
...
+> docker exec -it unnumbered-p2p-peer ip route
+5.5.5.5 nhid 21 proto bgp metric 20
+        nexthop via inet6 fe80::dcad:beff:feff:1160 dev eth00 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1161 dev eth01 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1162 dev eth02 weight 1
+```
+
+### The API Extension
+
+We have three option:
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:366]: to do what? Can we have a premise saying what we want in frr? Just a couple of sentences

File: design/unnumbered-bgp.md (hunk 15 above)
Comment (fedepaol) [reply design/unnumbered-bgp.md:376]: +1 on this

File: design/unnumbered-bgp.md [hunk 28]
```diff
@@ -0,0 +1,508 @@
...
+metadata:
+  name: peer-unnumber
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  ASN: 645000
+  interface: net0
+```
+
+#### Option 3. Add new Unnumber bool field
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:428]: Let's remove this, I think it's a bit confusing.

File: design/unnumbered-bgp.md [hunk 29]
```diff
@@ -0,0 +1,508 @@
...
+  name: peer-unnumber
+  namespace: metallb-system
+spec:
+  myASN: 64512
+  ASN: 645000
+  peerAddress: net0
+  unnumber: true
+```
+
+### Testing
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:469]:
```suggestion
### Test plan
```

File: design/unnumbered-bgp.md [hunk 30]
```diff
@@ -0,0 +1,508 @@
...
+
+#### E2E test for [Unnumber on primary interface](design/unnumbered-bgp.md#story-1-unnumbered-iebgp-on-primary-node-interface)
+
+There is no straightforward way to test unnumbered in the primary node
+interface. The testing setup (kind creation) creates bridge and connect
+containers. Ideally we need to setup a kind cluster, and replace the bridge
+with a containers that runs FRR on each other side of the veth, and at the same
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:476]:
```suggestion
with a container that runs FRR on each other side of the veth, and at the same
```

File: design/unnumbered-bgp.md [hunk 31]
```diff
@@ -0,0 +1,508 @@
...
+time to provide switch/upstream. Until we find a way to automate that infra setup
+we test that scenario manual.
+
+#### E2E test for [Unnumber on secondary interface](design/unnumbered-bgp.md#story-1-unnumbered-iebgp-on-secondary-node-interface)
+
+We should modify the testing infra to include point-to-point links which means
+connect containers using VETH, and not through a bridge. We need NET-ADMIN cap
+(or sudo) when creating the VETH.
+
+##### Option 1 - have a separate test
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:486]: I'd start with this, the majority of the scenarios are functional and should (!!!) be orthogonal to how we establish the session

File: design/unnumbered-bgp.md [hunk 32]
```diff
@@ -0,0 +1,508 @@
...
+
+### The API Extension
+
+We have three option:
+
+#### Option 1. Re-Use Address field
+
+We can re-use `Address string` to either be IP or interface, if not valid IP we
+indicate Unnumbered BGP peering. This looks the minimal API, but it will not
+have clean error path. E.g. if the user input is not valid IP, the
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:372]: done

File: design/unnumbered-bgp.md (hunk 15 above)
Comment (karampok) [reply design/unnumbered-bgp.md:376]: ACK then, will put that as but we need to decide if to further https://github.com/metallb/metallb/issues/2423
and have this 
````
// 1. if address and no interface then as normal
// 2. if non valid address and no interface then fail
// 3. if address is LLA and interface not empty, then LLA support
// == extra FRR directive `neighbor fe80::a876:4dff:fe77:408 interface eth0`
// 4. If address is LLA and empty interface, then fail
// 5. If address is empty and interface has value then the BGP unnumber peering config will take place
// ==  modify FRR directive `neighbor PEER interface remote-as <ASN>`
// && replace any address directive to use interface e.g.  `neighbor net0 timers 180 540`
// && enable RAs on this interface
// 6. If address has value (non-LLA) and interface non empty, then fail
``` 
as the end API

File: design/unnumbered-bgp.md (hunk 28 above)
Comment (karampok) [reply design/unnumbered-bgp.md:428]: This is my preference :), I will keep it a rejected option without code example

File: design/unnumbered-bgp.md (hunk 29 above)
Comment (karampok) [reply design/unnumbered-bgp.md:469]: done

File: design/unnumbered-bgp.md [hunk 33]
```diff
@@ -0,0 +1,508 @@
...
+spec:
+  myASN: 64512
+  ASN: 645000
+  peerAddress: net0
+  unnumber: true
+```
+
+### Testing
+
+#### E2E test for [Unnumber on primary interface](design/unnumbered-bgp.md#story-1-unnumbered-iebgp-on-primary-node-interface)
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:471]: done

File: design/unnumbered-bgp.md (hunk 30 above)
Comment (karampok) [reply design/unnumbered-bgp.md:476]: Done

File: design/unnumbered-bgp.md [hunk 34]
```diff
@@ -0,0 +1,508 @@
...
+
+There is no straightforward way to test unnumbered in the primary node
+interface. The testing setup (kind creation) creates bridge and connect
+containers. Ideally we need to setup a kind cluster, and replace the bridge
+with a containers that runs FRR on each other side of the veth, and at the same
+time to provide switch/upstream. Until we find a way to automate that infra setup
+we test that scenario manual.
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:478]: As discussed is not worth the effort to refactor how we set up the infra.

File: design/unnumbered-bgp.md [hunk 35]
```diff
@@ -0,0 +1,508 @@
...
+
+#### E2E test for [Unnumber on secondary interface](design/unnumbered-bgp.md#story-1-unnumbered-iebgp-on-secondary-node-interface)
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:480]: done

File: design/unnumbered-bgp.md (hunk 31 above)
Comment (karampok) [reply design/unnumbered-bgp.md:486]: Sounds good, i keep the text as is, one we have option 1 I would say we can merge, but I am in favor to then try to implement also option

File: design/unnumbered-bgp.md [hunk 36]
```diff
@@ -0,0 +1,508 @@
...
+
+Displayed  2 routes and 4 total paths
+> docker exec -it unnumbered-p2p-peer ip route
+5.5.5.5 nhid 21 proto bgp metric 20
+        nexthop via inet6 fe80::dcad:beff:feff:1160 dev eth00 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1161 dev eth01 weight 1
+        nexthop via inet6 fe80::dcad:beff:feff:1162 dev eth02 weight 1
+```
+
+### The API Extension
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:364]: done

File: design/unnumbered-bgp.md (hunk 27 above)
Comment (karampok) [reply design/unnumbered-bgp.md:366]: Three options how the API could look like, but that was removed.

File: design/unnumbered-bgp.md (hunk 13 above)
Comment (karampok) [reply design/unnumbered-bgp.md:331]: I was expected if you could explain me how the RouterID is important or not. Example when we have single IPv6 stack. But I have that for discussion, removed

File: design/unnumbered-bgp.md [hunk 37]
```diff
@@ -0,0 +1,508 @@
...
+Neighbor                  V         AS   MsgRcvd   MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd   PfxSnt Desc
+unnumbered-p2p-peer(net0) 4      65004       512       512        1    0    0 00:25:15            0        1 TOR
+
+Total number of neighbors 1
+```
+### Topics
+
+#### Loopback IPv4 address
+
+Linux sends and receives IPv4 packets over an interface that does not have an IPv4 address.
```
[the remark this answers is not in the document]
Comment (karampok) [reply design/unnumbered-bgp.md:318]: TODO, explain better

**Review by karampok — COMMENTED** [review]
thanks for the feedback

**Comment by karampok** [comment]
@fedepaol  @oribon I have updated the doc and I think I have addressed all your feedback. Please take a look, thanks

File: design/unnumbered-bgp.md (hunk 15 above)
Comment (oribon) [reply design/unnumbered-bgp.md]: not sure I understand - implementation wise the issue will be solved for free(?) and documentation wise it seems sufficient to document the `Interface` field something like "if specified without address then peering happens over the interface (unnumbered), if together with address then it is assumed to be LLA"?

File: design/unnumbered-bgp.md [hunk 38]
```diff
@@ -0,0 +1,541 @@
...
+  are the same across the cluster nodes)
+- Remove the task of network admins to setup up addressing on those interfaces
+  if it is needed only for BGP peering
+
+## Non Goals
+
+- Make it work over no point-to-point connection
+- Define/Discuss dual ToR architecture, on top of bond interfaces or other HA strategies
+- [Dynamic BGP Peering](https://ipwithease.com/dynamic-bgp-peering/) is a
+  feature different than unnumbered BGP and not supported by Metallb. Both
```
Comment (oribon) [inline design/unnumbered-bgp.md:41]: nit: remove the "not supported by MetalLB" (it might be in the future, and unrelated to this)

File: design/unnumbered-bgp.md [hunk 39]
```diff
@@ -0,0 +1,541 @@
...
+  feature reduce the configuration needed for the BGP peering. The main difference is
+  that Dynamic BGP peering still needs IP addresses and does not require point-to-point links.
+- Support `remote-as` to accept `<external|internal|auto>`
```
Comment (oribon) [inline design/unnumbered-bgp.md:44]: I'd also remove this (seems unrelated)

File: design/unnumbered-bgp.md [hunk 40]
```diff
@@ -0,0 +1,541 @@
...
+As network administrator, I want to reduce the size of the configuration in the ToR router.
+
+## The API Changes
+
+### Option 2. Add a new Interface field
+
+We can introduce a new field `Interface string` which if defined then Unnumber
+BGP takes place. In that case, the address must NOT have value. For
+completeness, we could use field might be used to support `LLA%zone` [user case](https://github.com/metallb/metallb/issues/2423).
+So at the end, the API documentation will look like
```
Comment (oribon) [inline design/unnumbered-bgp.md:68]: as the other comment, I think this could be a lot shorter (no need to specify everything and its consequence)

File: design/unnumbered-bgp.md (hunk 39 above)
Comment (karampok) [reply design/unnumbered-bgp.md]: done

File: design/unnumbered-bgp.md (hunk 40 above)
Comment (karampok) [reply design/unnumbered-bgp.md:67]: I agree, but I prefer to remove once @fedepaol take another feedback and he is ok with the consequence or not (we could decide no to support the LLA)

Comment (fedepaol) [reply design/unnumbered-bgp.md:67]: Let's focus on unnumbered and maybe implement LLA afterwards if we need / want to.
What's important here is that the API is compatible with that. If we want to mention it, let's add a subsection in the bottom

File: design/unnumbered-bgp.md [hunk 41]
```diff
@@ -0,0 +1,540 @@
...
+
+### Story 3
+
+As network administrator, I want to reduce the size of the configuration in the ToR router.
+
+## The API Changes
+
+### Option 2. Add a new Interface field
+
+We can introduce a new field `Interface string` which if defined then Unnumber
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:64]: Since this is the proposed change, let's reword with `A new interface field will be introduced`

File: design/unnumbered-bgp.md [hunk 42]
```diff
@@ -0,0 +1,540 @@
...
+As network administrator, I want to avoid configuring IP addresses for
+point-to-point connection just for the BPG peering reason.
+
+### Story 3
+
+As network administrator, I want to reduce the size of the configuration in the ToR router.
+
+## The API Changes
+
+### Option 2. Add a new Interface field
```
Comment (fedepaol) [inline design/unnumbered-bgp.md:62]: This is not an option anymore, but the proposed change

Comment (karampok) [reply design/unnumbered-bgp.md:62]: done

File: design/unnumbered-bgp.md (hunk 41 above)
Comment (karampok) [reply design/unnumbered-bgp.md]: done

File: design/unnumbered-bgp.md (hunk 40 above)
Comment (karampok) [reply design/unnumbered-bgp.md]: ok, removed the reference to LLA

**Comment by fedepaol** [comment]
LGTM, thanks for this!
